crates/sui-framework/sources/governance/genesis.move

```diff
@@ -4,128 +4,206 @@
-            let net_address = *vector::borrow(&validator_net_addresses, i);
-            let p2p_address = *vector::borrow(&validator_p2p_addresses, i);
-            let primary_address = *vector::borrow(&validator_primary_addresses, i);
-            let worker_address = *vector::borrow(&validator_worker_addresses, i);
-            let gas_price = *vector::borrow(&validator_gas_prices, i);
-            let commission_rate = *vector::borrow(&validator_commission_rates, i);
-            vector::push_back(&mut validators, validator::new(
+            let GenesisValidatorMetadata {
+                name,
+                description,
+                image_url,
+                project_url,
+                sui_address,
+                gas_price,
+                commission_rate,
+                protocol_public_key,
+                proof_of_possession,
+                network_public_key,
+                worker_public_key,
+                network_address,
+                p2p_address,
+                primary_address,
+                worker_address,
+            } = *vector::borrow(&genesis_validators, i);
+
+            let validator = validator::new(
                 sui_address,
-                pubkey,
-                network_pubkey,
-                worker_pubkey,
+                protocol_public_key,
+                network_public_key,
+                worker_public_key,
                 proof_of_possession,
                 name,
                 description,
                 image_url,
                 project_url,
-                net_address,
+                network_address,
                 p2p_address,
                 primary_address,
                 worker_address,
-                // Initialize all validators with uniform stake taken from the subsidy fund.
-                // TODO: change this back to take from subsidy fund instead.
-                option::some(balance::split(&mut sui_supply, initial_validator_stake_mist)),
+                // Initialize all validators with no stake. Stake distribution
+                // will be handled by the 'allocate_tokens' call bellow.
+                option::none(),
                 gas_price,
                 commission_rate,
-                true, // validator is active right away
                 ctx
-            ));
+            );
+
+            vector::push_back(&mut validators, validator);
+
             i = i + 1;
         };
 
+        // Allocate tokens and staking operations
+        allocate_tokens(
+            sui_supply,
+            token_distribution_schedule.allocations,
+            &mut validators,
+            ctx
+        );
+
+        // Activate all validators
+        activate_validators(&mut validators);
+
         sui_system::create(
             validators,
             subsidy_fund,
             storage_fund,
-            governance_start_epoch,
+            genesis_chain_parameters.governance_start_epoch,
             INIT_STAKE_SUBSIDY_AMOUNT,
```

Comment:
It would make sense to put this into the chain parameters too?

Reply:
Yep, planning on doing that in my follow up, plus the stake departure thresholds as well.